### DISCUSSION TITLE
Make Curve resource editable in the Animation window's BezierCurve editor

### DISCUSSION THREAD

**Discussion by Kartopod** [godotengine/godot-proposals#11455] [Editor]
In my opinion the UI in the inspector for editing curves is completely unusable. We already have a dedicated editor to edit BezierCurveTracks in the AnimationPlayer's 'Animation' window, which is far superior in terms of UX and usability. Allowing editing curves with this editor would improve the UX when working with curves tremendously. 

To clarify, when I say **Curve resource editor**, I am referring to this UI in the inspector:
![image](https://github.com/user-attachments/assets/a54e57f9-1502-4e59-b7e6-998871627960)
and when I say BezierCurve editor, I am referring to the animation window's curve editor. 
![image](https://github.com/user-attachments/assets/26019589-081e-4344-8817-c3ba960593d0)

### Cons of the current implementation of the curve resource editor:
- The curve is tiny and you cannot zoom in or open the graph in a separate window to make it larger. This makes it difficult to make precise changes to values. The only way to make the graph larger is to make the entire inspector window larger, which is not ideal.

- Cannot edit multiple keys at once. 
![Edit multiple keys](https://github.com/user-attachments/assets/01ddd6c9-901c-4620-9c65-b15ce6d8d296)

- Can only effect the rotation of the handles, but not their length. Of course, you can work around this by adding more keyframes, but it is still not intuitive and that further clutters the already tiny graph. 
![Resize Handles](https://github.com/user-attachments/assets/b3d3aded-1328-475d-943d-49fb069cb476)

- In addition to the previous point, you only have the ability to set the handle mode to 'Linear' and 'Free', but not 'Balanced' or 'Mirrored', which are both extremely useful when editing curves. 
![Mirrored and Balanced](https://github.com/user-attachments/assets/d63534d7-404e-468c-9b2d-6bf7eda85c2c)

---
All of these features can be utilized by just reusing already existing functionality of the BezierCurve editor, so I don't see why it shouldn't be reused for editing the curve resource. In my opinion, the BezierCurve editor's UX also suboptimal and has a long way to go, but the curve resource editor makes it look like a godsend when compared. I am honestly surprised to not find any complaints about the existing curve resource editor.

**Comment by Kartopod** [comment]
I was very excited about https://github.com/godotengine/godot/pull/82306, however in practice the UX to edit curves makes me want to avoid it completely. 

Fortunately, I was able to figure out a workaround to be able to use the BezierCurve editor with the `set_custom_interpolator()`. You can do this by creating a dummy AnimationPlayer node, create a BezierCurveTrack with a random float property on any node and edit the curve to your liking. When done click on 'Manage animations...' under the 'Animation' button on top. Here, you can save the animation to the FileSystem as an `Animation` resource.

You can then export a reference to the `Animation` you created and use `Animation.bezier_track_interpolate()` to sample the value from the curve at a given time, similar to how you would do it with `Curve.sample_baked()`, and use it in tandem with `PropertyTweener.set_custom_interpolator()`


https://github.com/user-attachments/assets/ecfea307-4c24-4f1f-813d-c5ddfc2ba079

**Comment by BaconEggsRL** [comment]
Any progress on this proposal?
Being able to drag the handles length would be a big improvement.

From the docs it makes it seem like we should be able to define arbitrary control point positions, building off of the cubic bezier:
https://docs.godotengine.org/en/stable/tutorials/math/beziers_and_curves.html#adding-control-points

I appreciate the animation track workaround, but having an editable exported resource is quite handy for quick changes.

Looking for something like in CSS, where you can define cubic-bezier easily for easing functions:
https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/easing-function/cubic-bezier
https://cubic-bezier.com/#.33,1,.68,1

It would be useful to have all the easing functions as curve presets in Godot:
https://easings.net/

**Comment by pixequil** [comment]
Kartopod's method saved my life, I'm glad there's an option for this. But it feels like such a hacky workaround. Like, this is some stone age stuff here. This has got to be integrated into the real interface as an intended thing people can do.
